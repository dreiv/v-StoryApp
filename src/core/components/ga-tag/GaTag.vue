<script setup lang="ts">
import { computed, useCssModule, useTemplateRef } from 'vue'

export interface TagProps {
  label?: string
  beforeIcon?: string | object
  afterIcon?: string | object
  separator?: boolean
  interactive?: boolean
  selected?: boolean
  success?: boolean
  information?: boolean
  error?: boolean
  warning?: boolean
  disabled?: boolean
  utility?:
    | 'grey'
    | 'indigo'
    | 'teal'
    | 'lime'
    | 'purple'
    | 'pink'
    | 'rose'
    | 'yellow'
    | 'blue'
    | 'green'
    | 'orange'
    | 'disabled'
}

const emit = defineEmits(['onDelete'])
const {
  afterIcon,
  disabled,
  error,
  information,
  interactive,
  selected,
  success,
  utility,
  warning,
} = defineProps<TagProps>()
const style = useCssModule()

const componentRef = useTemplateRef<HTMLElement>('componentRef')

const classes = computed(() => {
  const classList = [style.tag]

  if (utility) {
    classList.push(style.utility, style[utility])
  }

  if (interactive) {
    classList.push(style.interactive)
    if (selected) classList.push(style.selected)
  }

  if (disabled) {
    classList.push(style.disabled)
  } else if (information) {
    classList.push(style.information)
  } else if (error) {
    classList.push(style.error)
  } else if (warning) {
    classList.push(style.warning)
  } else if (success) {
    classList.push(style.success)
  } else {
    classList.push(style.default)
  }

  return classList
})

const is = computed(() => (interactive || afterIcon ? 'button' : 'span'))

const handleKeyup = (event: KeyboardEvent) => {
  if (disabled) return
  if (event.key === 'Backspace' || event.key === 'Delete') {
    emit('onDelete')
  } else if (event.key === 'Escape' && componentRef.value) {
    componentRef.value.blur()
  }
}
</script>

<template>
  <component :is :class="classes" :disabled @keyup="handleKeyup" ref="componentRef">
    <slot name="before-icon">
      <component :is="beforeIcon" v-if="beforeIcon" :class="$style.beforeIcon" :size="14" />
    </slot>
    <slot>{{ label }}</slot>

    <span v-if="separator" :class="$style.separator" />
    <slot name="after-icon">
      <component :is="afterIcon" v-if="afterIcon" :class="$style.afterIcon" :size="14" />
    </slot>
  </component>
</template>

<style module>
.tag {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  gap: var(--ga-size-spacing-02);

  border-radius: var(--ga-radius-round);
  padding: var(--ga-size-spacing-02) var(--ga-size-spacing-03);

  font-size: var(--ga-text-sm-font-size);
  letter-spacing: var(--ga-text-sm-tracking);

  &:focus-visible {
    outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
    outline-offset: var(--ga-size-spacing-01);
  }

  &:enabled:is(:hover, :focus) {
    .afterIcon {
      cursor: pointer;
      color: var(--ga-color-icon-action-hover);
    }
  }
}

.default {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-action);
  background-color: var(--ga-color-surface-primary);
  color: var(--ga-color-text-action);
}

.separator {
  background-color: var(--ga-color-border-disabled);
  width: 1px;
  height: 100%;
}

.success {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-success);
  background-color: var(--ga-color-utility-green-light);
  color: var(--ga-color-text-success);

  .beforeIcon {
    color: var(--ga-color-icon-success);
  }
}

.interactive {
  &:not(.selected, .utility) {
    border-style: dashed;
  }

  &:enabled {
    cursor: pointer;
  }

  &:not(:active, :disabled):is(:hover, :focus) {
    background-color: var(--ga-color-surface-action-hover-2);

    .beforeIcon,
    .afterIcon {
      color: var(--ga-color-icon-action-hover);
    }
  }

  &:disabled {
    cursor: not-allowed;
    border-color: var(--ga-color-surface-disable-selected);
    background-color: var(--ga-color-surface-disable-selected);
    color: var(--ga-color-text-on-action);

    .beforeIcon,
    .afterIcon {
      color: var(--ga-color-icon-on-primary);
    }

    &:not(.selected) {
      border-color: var(--ga-color-border-disabled);
      background-color: var(--ga-color-surface-primary);
      color: var(--ga-color-text-disabled);

      .beforeIcon,
      .afterIcon {
        color: var(--ga-color-icon-disabled);
      }
    }
  }
}

.information {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-information);
  background-color: var(--ga-color-utility-blue-light);
  color: var(--ga-color-text-information);

  .beforeIcon {
    color: var(--ga-color-icon-information);
  }
}

.error {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-error);
  background-color: var(--ga-color-utility-red-light);
  color: var(--ga-color-text-error);

  .beforeIcon {
    color: var(--ga-color-icon-error);
  }
}

.warning {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-warning);
  background-color: var(--ga-color-utility-orange-light);
  color: var(--ga-color-text-warning);

  .beforeIcon {
    color: var(--ga-color-icon-warning);
  }
}

.disabled {
  border: var(--ga-size-border-width-sm) solid var(--ga-color-border-disabled);
  background-color: var(--ga-color-surface-disabled);
  color: var(--ga-color-text-disable-selected);

  .beforeIcon {
    color: var(--ga-color-icon-on-disabled);
  }
}

.utility {
  border-radius: var(--ga-radius);

  .beforeIcon {
    color: var(--ga-color-text-body);
  }

  &.grey {
    border-color: var(--ga-color-utility-grey);
    background-color: var(--ga-color-utility-grey-light);
  }

  &.indigo {
    border-color: var(--ga-color-utility-indigo);
    background-color: var(--ga-color-utility-indigo-light);
  }

  &.teal {
    border-color: var(--ga-color-utility-teal);
    background-color: var(--ga-color-utility-teal-light);
  }

  &.lime {
    border-color: var(--ga-color-utility-lime);
    background-color: var(--ga-color-utility-lime-light);
  }

  &.purple {
    border-color: var(--ga-color-utility-purple);
    background-color: var(--ga-color-utility-purple-light);
  }

  &.pink {
    border-color: var(--ga-color-utility-pink);
    background-color: var(--ga-color-utility-pink-light);
  }

  &.rose {
    border-color: var(--ga-color-utility-rose);
    background-color: var(--ga-color-utility-rose-light);
  }

  &.yellow {
    border-color: var(--ga-color-utility-yellow);
    background-color: var(--ga-color-utility-yellow-light);
  }

  &.blue {
    border-color: var(--ga-color-utility-blue);
    background-color: var(--ga-color-utility-blue-light);
  }

  &.green {
    border-color: var(--ga-color-utility-green);
    background-color: var(--ga-color-utility-green-light);
  }

  &.orange {
    border-color: var(--ga-color-utility-orange);
    background-color: var(--ga-color-utility-orange-light);
  }
}
</style>
