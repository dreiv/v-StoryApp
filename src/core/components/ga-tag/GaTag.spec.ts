import { mount, shallowMount, type VueWrapper } from '@vue/test-utils'
import GaTag from './GaTag.vue'

const MockIcon = {
  template: '<div class="mock-icon">Icon</div>',
  props: ['size'],
}

describe('GaTag', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = shallowMount(GaTag)
  })

  it('should render correctly with default props', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with label prop', () => {
    wrapper = shallowMount(GaTag, { props: { label: 'Test Tag' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Test Tag')
  })

  it('should render correctly with default slot content', () => {
    wrapper = shallowMount(GaTag, { slots: { default: 'Custom Label' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Custom Label')
  })

  it('should render beforeIcon when prop is provided', () => {
    wrapper = shallowMount(GaTag, { props: { beforeIcon: MockIcon } })
    expect(wrapper.find('.beforeIcon').exists()).toBe(true)
  })

  it('should render custom before-icon slot', () => {
    wrapper = shallowMount(GaTag, {
      slots: { 'before-icon': '<div class="custom-before-icon">Before</div>' },
    })
    expect(wrapper.find('.custom-before-icon').exists()).toBe(true)
  })

  it('should render afterIcon when prop is provided', () => {
    wrapper = shallowMount(GaTag, { props: { afterIcon: MockIcon } })
    expect(wrapper.find('.afterIcon').exists()).toBe(true)
  })

  it('should render custom after-icon slot', () => {
    wrapper = shallowMount(GaTag, {
      slots: { 'after-icon': '<div class="custom-after-icon">After</div>' },
    })
    expect(wrapper.find('.custom-after-icon').exists()).toBe(true)
  })

  it('should apply the default class when no variant props are set', () => {
    expect(wrapper.find('.default').exists()).toBe(true)
  })

  it.each(['information', 'error', 'warning', 'success', 'disabled'])(
    'should apply the %s class when the corresponding prop is set to true',
    async (variant) => {
      await wrapper.setProps({ [variant]: true })

      expect(wrapper.find(`.${variant}`).exists()).toBe(true)
      expect(wrapper.find('.default').exists()).toBe(false)
    },
  )

  it('should apply interactive class when interactive prop is true', async () => {
    await wrapper.setProps({ interactive: true })
    expect(wrapper.find('.interactive').exists()).toBe(true)
  })

  it('should apply selected class when interactive and selected props are true', async () => {
    await wrapper.setProps({ interactive: true, selected: true })

    expect(wrapper.find('.interactive').exists()).toBe(true)
    expect(wrapper.find('.selected').exists()).toBe(true)
  })

  it.each([
    'grey',
    'indigo',
    'teal',
    'lime',
    'purple',
    'pink',
    'rose',
    'yellow',
    'blue',
    'green',
    'orange',
  ])('should apply the utility and %s classes when utility prop is set to true', async (color) => {
    await wrapper.setProps({ utility: color })
    expect(wrapper.find('.utility').exists()).toBe(true)
    expect(wrapper.find(`.${color}`).exists()).toBe(true)
  })

  it('should render separator when separator prop is true', async () => {
    await wrapper.setProps({ separator: true })
    expect(wrapper.find('.separator').exists()).toBe(true)
  })

  it('should not render separator when separator prop is false', () => {
    expect(wrapper.find('.separator').exists()).toBe(false)
  })

  it.each([
    ['disabled', 'error'],
    ['disabled', 'warning'],
    ['disabled', 'success'],
    ['disabled', 'information'],
    ['information', 'error'],
    ['information', 'warning'],
    ['information', 'success'],
    ['error', 'warning'],
    ['error', 'success'],
    ['warning', 'success'],
  ])(
    'should prioritize %s over %s when both are set',
    async (priorityVariant, secondaryVariant) => {
      await wrapper.setProps({
        [priorityVariant]: true,
        [secondaryVariant]: true,
      })
      expect(wrapper.find(`.${priorityVariant}`).exists()).toBe(true)
      expect(wrapper.find(`.${secondaryVariant}`).exists()).toBe(false)
    },
  )

  it('should render as a button when interactive is true', async () => {
    await wrapper.setProps({ interactive: true })
    expect(wrapper.element.tagName).toBe('BUTTON')
  })

  it('should render as a button when afterIcon is provided', async () => {
    await wrapper.setProps({ afterIcon: MockIcon })
    expect(wrapper.element.tagName).toBe('BUTTON')
  })

  it('should render as a span by default', () => {
    expect(wrapper.element.tagName).toBe('SPAN')
  })

  // Event tests
  it('should emit onDelete event when afterIcon is clicked', async () => {
    wrapper = mount(GaTag, { props: { afterIcon: MockIcon } })
    await wrapper.findComponent('.afterIcon').trigger('click')
    expect(wrapper.emitted('onDelete')).toBeTruthy()
    expect(wrapper.emitted('onDelete')?.length).toBe(1)
  })

  it('should emit onDelete event when Delete key is pressed', async () => {
    await wrapper.trigger('keyup', { key: 'Delete' })
    expect(wrapper.emitted('onDelete')).toBeTruthy()
    expect(wrapper.emitted('onDelete')?.length).toBe(1)
  })

  it('should emit onDelete event when Backspace key is pressed', async () => {
    await wrapper.trigger('keyup', { key: 'Backspace' })
    expect(wrapper.emitted('onDelete')).toBeTruthy()
    expect(wrapper.emitted('onDelete')?.length).toBe(1)
  })

  it('should not emit onDelete event when other keys are pressed', async () => {
    await wrapper.trigger('keyup', { key: 'Enter' })
    expect(wrapper.emitted('onDelete')).toBeFalsy()
  })

  it('should not emit onDelete event when disabled', async () => {
    await wrapper.setProps({ disabled: true })
    await wrapper.trigger('keyup', { key: 'Delete' })
    expect(wrapper.emitted('onDelete')).toBeFalsy()
  })

  it('should call blur on Escape key', async () => {
    const blurSpy = vi.spyOn(HTMLElement.prototype, 'blur').mockImplementation(() => {})
    const mountedWrapper = mount(GaTag)

    mountedWrapper.vm.$refs.componentRef = mountedWrapper.element

    await mountedWrapper.trigger('keyup', { key: 'Escape' })
    expect(blurSpy).toHaveBeenCalled()

    blurSpy.mockRestore()
  })
})
