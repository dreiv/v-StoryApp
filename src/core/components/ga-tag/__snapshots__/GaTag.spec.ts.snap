// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GaTag > should render correctly with default props 1`] = `"<span class="tag default" disabled="false"><!--v-if--><!--v-if--><!--v-if--></span>"`;

exports[`GaTag > should render correctly with default slot content 1`] = `"<span class="tag default" disabled="false"><!--v-if-->Custom Label<!--v-if--><!--v-if--></span>"`;

exports[`GaTag > should render correctly with label prop 1`] = `"<span class="tag default" disabled="false"><!--v-if-->Test Tag<!--v-if--><!--v-if--></span>"`;
